src/view.c

```diff
@@ -18,6 +18,80 @@
 #define LAB_FALLBACK_WIDTH  640
 #define LAB_FALLBACK_HEIGHT 480
 
+static bool
+matches_criteria(struct view *view, enum lab_view_criteria criteria)
+{
+	if (!isfocusable(view)) {
+		return false;
+	}
+	if (criteria & LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		/*
+		 * Always-on-top views are always on the current desktop and are
+		 * special in that they live in a different tree.
+		 */
+		struct server *server = view->server;
+		if (view->scene_tree->node.parent != server->workspace_current->tree
+				&& !view_is_always_on_top(view)) {
+			return false;
+		}
+	}
+	if (criteria & LAB_VIEW_CRITERIA_NO_ALWAYS_ON_TOP) {
+		if (view_is_always_on_top(view)) {
+			return false;
+		}
+	}
+	if (criteria & LAB_VIEW_CRITERIA_NO_SKIP_WINDOW_SWITCHER) {
+		if (window_rules_get_property(view, "skipWindowSwitcher") == LAB_PROP_TRUE) {
+			return false;
+		}
+	}
+	return true;
+}
+
+struct view *
+view_first(struct wl_list *head, enum lab_view_criteria criteria)
+{
+	struct view *view = wl_container_of(head->next, view, link);
```

Comment:
AFAIR in an empty list, `head->next` would be `head` itself. Thus its not actually a view but `server->views` in this case. So I think we need something like `if (head->next == head) { return NULL; }` as first statement here.

The reason this didn't actually bit us here was that we currently still return early in `osd_update()` on `wl_list_empty(node_list))` with `node_list` being the current workspace scene_graph child nodes (or on `!server->cycle_view`). Thus this function was never called with an empty view list.

Reply:
Good point. Returning a view didn't feel right! Have implemented.

Have tested with the code below to prove that we deal with an empty list properly.

```
diff --git a/src/action.c b/src/action.c
index c04e1c5..cd9a35d 100644
--- a/src/action.c
+++ b/src/action.c
@@ -557,7 +557,10 @@ actions_run(struct view *activator, struct server *server,
 			}
 			break;
 		case ACTION_TYPE_DEBUG:
-			debug_dump_scene(server);
+			for_each_view(view, &server->views, 0) {
+				wlr_log(WLR_ERROR, "id=%s", view_get_string_prop(view, "app_id"));
+			}
+			// debug_dump_scene(server);
 			break;
 		case ACTION_TYPE_EXECUTE:
 			{
```